### ISSUE TITLE
Compiler allows mutating immutable variables in unreachable code

### ISSUE THREAD

**Issue by Diggsey** [rust-lang/rust#49788]
IMO, this should give a compiler error:
https://play.rust-lang.org/?gist=13095cf519ab250005688bf214e2482f&version=stable

```rust
fn main() {
    let x = 1;
    return;
    x += 2;
}
```

**Comment by Spoonbender** [comment]
Triage: no change
